### PR TITLE
Render plain text README.txt monospaced

### LINKED ISSUE #2718: render plain text README.txt with monospace font
- Gitea version:
1.2.1
- Example
https://try.gitea.io/pmorjan/issue2701
## Description
A plain text README.txt should be rendered on the repo home page with a
 monospace font to  preserve text formatting and indentations.

### PR DESCRIPTION
**Pull request by jonasfranz**
Fixes #2718 

Will add `<pre>TEXT</pre>` at the readme page.

## Screenshot
screenshot

### CODE REVIEW

**Comment by Morlinest** [comment]
This will break markdown rendering.

**Comment by jonasfranz** [comment]
@Morlinest Why? It is not used when markup is rendered. (See acaf4b3)

**Comment by Morlinest** [comment]
@JonasFranzDEV I was checking your first version (commit). Will look at it again.

File: routers/repo/view.go [hunk 1]
```diff
@@ -195,9 +195,11 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
 		ctx.Data["ReadmeExist"] = readmeExist
 		if markup.Type(blob.Name()) != "" {
 			ctx.Data["IsRenderedHTML"] = true
+			ctx.Data["IsMarkup"] = true
 			ctx.Data["FileContent"] = string(markup.Render(blob.Name(), buf, path.Dir(treeLink), ctx.Repo.Repository.ComposeMetas()))
 		} else if readmeExist {
 			ctx.Data["IsRenderedHTML"] = true
+			ctx.Data["IsMarkup"] = false
```
Comment (Morlinest) [inline routers/repo/view.go]: You can keep only `ctx.Data["IsMarkup"] = true`

File: templates/repo/view_file.tmpl [hunk 2]
```diff
@@ -38,7 +38,11 @@
 	<div class="ui attached table segment">
 		<div class="file-view {{if .IsRenderedHTML}}markdown{{else if .IsTextFile}}code-view{{end}} has-emoji">
 			{{if .IsRenderedHTML}}
-				<pre>{{if .FileContent}}{{.FileContent | Str2html}}{{end}}</pre>
+				{{if not .IsMarkup}}
```
Comment (Morlinest) [inline templates/repo/view_file.tmpl]: Switch order, test for `{{if .IsMarkup}}`

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: templates/repo/view_file.tmpl [hunk 3]
```diff
@@ -38,7 +38,11 @@
 	<div class="ui attached table segment">
 		<div class="file-view {{if .IsRenderedHTML}}markdown{{else if .IsTextFile}}code-view{{end}} has-emoji">
 			{{if .IsRenderedHTML}}
-				{{if .FileContent}}{{.FileContent | Str2html}}{{end}}
+				{{if .IsMarkup}}
+					{{if .FileContent}}{{.FileContent | Str2html}}{{end}}
+				{{else}}
+					<pre>{{if .FileContent}}{{.FileContent | Str2html}}{{end}}</pre>
```
Comment (Morlinest) [inline templates/repo/view_file.tmpl]: Is rly `Str2html` needed here?
`<pre>` should be inside `{{if .FileContent}}` block.

Edit: Probably it's safe to always use `Str2html`

File: routers/repo/view.go [hunk 4]
```diff
@@ -195,6 +195,7 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
 		ctx.Data["ReadmeExist"] = readmeExist
 		if markup.Type(blob.Name()) != "" {
 			ctx.Data["IsRenderedHTML"] = true
+			ctx.Data["IsMarkup"] = true
```
Comment (Morlinest) [inline routers/repo/view.go]: Please look for other places with `markup.Type` check. I think it (`ctx.Data["IsMarkup"] = true`) should be added also there (I found https://github.com/JonasFranzDEV/gitea/blob/5aca5a8a3d618630d10272445badcbea5381683f/routers/repo/view.go#L97).

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: templates/repo/view_file.tmpl [hunk 5]
```diff
@@ -38,7 +38,11 @@
 	<div class="ui attached table segment">
 		<div class="file-view {{if .IsRenderedHTML}}markdown{{else if .IsTextFile}}code-view{{end}} has-emoji">
```
Comment (lafriks) [inline templates/repo/view_file.tmpl]: Please add here `{{.IsMarkup}}markdown{{else}}plain-text{{end}}` in place of `markdown` and in `_repository.less` file after line 271 add style:
```css
			.plain-text {
				padding: 1em 2em 1em 2em;
			}
```

I think this way looks much better:
attels

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/repo/view.go [hunk 6]
```diff
@@ -195,6 +195,7 @@ func renderFile(ctx *context.Context, entry *git.TreeEntry, treeLink, rawLink st
 		ctx.Data["ReadmeExist"] = readmeExist
 		if markup.Type(blob.Name()) != "" {
 			ctx.Data["IsRenderedHTML"] = true
```
Comment (Morlinest) [inline routers/repo/view.go]: `ctx.Data["IsRenderedHTML"] = true` can be removed (check other places, same as `markup.Type`) if `ctx.Data["IsMarkup"] = true` is set, then you can remove nested `if` from template  and use `else if`.

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: templates/repo/view_file.tmpl (hunk 5 above)
Comment (jonasfranz) [reply templates/repo/view_file.tmpl]: This looks actually like this:


screenshot2

Is this still okay? I've left the `<pre>` stuff in place.

**EDIT:**
It looks fine Chrome (I've used safari). Maybe it is a cache problem in Safari.

Comment (lafriks) [reply templates/repo/view_file.tmpl]: Should be fine in all, try cleaning cache (also did you do `make generate-stylesheets`?)

File: routers/repo/view.go (hunk 6 above)
Comment (jonasfranz) [reply routers/repo/view.go]: Where should I remove the `if` in template? (Could you provide a Line?)

File: routers/repo/view.go (hunk 4 above)
Comment (jonasfranz) [reply routers/repo/view.go]: Done

File: templates/repo/view_file.tmpl (hunk 5 above)
Comment (jonasfranz) [reply templates/repo/view_file.tmpl]: > also did you do make generate-stylesheets?

Yes

File: routers/repo/view.go (hunk 6 above)
Comment (Morlinest) [reply routers/repo/view.go]: https://github.com/go-gitea/gitea/pull/2721/files#diff-2ec327528ed2be8dd3c2201ef8e6a42cR40

File: public/less/_repository.less [hunk 7]
```diff
@@ -270,6 +270,10 @@
 				}
 			}
 
+			.plain-text {
+            	padding: 1em 2em 1em 2em;
```
Comment (lafriks) [inline public/less/_repository.less]: Something looks wrong with spacing

File: routers/repo/view.go [hunk 8]
```diff
@@ -95,6 +95,7 @@ func renderDirectory(ctx *context.Context, treeLink string) {
 			buf = append(buf, d...)
 			ctx.Data["IsRenderedHTML"] = true
```
Comment (Morlinest) [inline routers/repo/view.go]: Move this line to line 101

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
LGTM

File: templates/repo/view_file.tmpl [hunk 9]
```diff
@@ -36,9 +36,11 @@
 		{{end}}
 	</h4>
 	<div class="ui attached table segment">
-		<div class="file-view {{if .IsRenderedHTML}}markdown{{else if .IsTextFile}}code-view{{end}} has-emoji">
-			{{if .IsRenderedHTML}}
+		<div class="file-view {{if .IsMarkup}}markdown{{else if .IsRenderedHTML}}plain-text{{else if .IsTextFile}}code-view{{end}} has-emoji">
+			{{if .IsMarkup}}
 				{{if .FileContent}}{{.FileContent | Str2html}}{{end}}
+			{{else if .IsRenderedHTML}}
+				<pre>{{if .FileContent}}{{.FileContent | Str2html}}{{end}}</pre>
```
Comment (Morlinest) [inline templates/repo/view_file.tmpl:43]: Move `<pre>` inside `{{if .FileContent}}` block. Don't need to render empty tag.

**Review by Morlinest — CHANGES_REQUESTED** [review]

File: templates/repo/view_file.tmpl (hunk 9 above)
Comment (lafriks) [reply templates/repo/view_file.tmpl:43]: @Morlinest it is better to render it because of margins as css is accounted for plain-text to have pre inside

Comment (Morlinest) [reply templates/repo/view_file.tmpl:43]: @lafriks OK then

**Review by Morlinest — APPROVED** [review]

**Comment by Morlinest** [comment]
LGTM

**Pull request merged**